Homepage search form

laurenmarietta opened the pull request:
Introduces new form on the web app homepage that allows users to immediately search for JWST proposals or rootnames (e.g. 86600; jw86600006001_02101_00008_guider1) and easily get to the appropriate `thumbnails` or `view_image` webpage. Changes include:
- Modifying `views.home` to handle the new search form
- Adding a `forms.py` file in the web app to create and validate the search form using Django's awesome object-oriented form method
- Adding a `fileroot_parser` util to `utils.py` for convenience

I threw this together kind of quickly, with much copy-and-pasting from StackOverflow, so please let me know if anything is particularly unclear or where there are any style guide violations!

Sadly, this PR does not achieve @bourque 's bonus of accepting a list of fileroots - that is going to require heavy alteration to `data_containers.thumbnails` and I think should be left for another time (or more adventurous soul?).

Closes #141.

bourque commented:
Thanks @laurenmarietta, I just tested this out locally and it works great!  I will try to have a look at the code this week.

laurenmarietta commented on jwql/website/apps/jwql/templates/home.html:
```diff
@@ -32,17 +35,35 @@ <h2>Welcome to the JWQL application.</h2>
 			</div>
 		</div>
 
-        <div class="row my-2" style="margin-left: 7rem; margin-right: 7rem;">
```
Okay, good to know you don't think this was premature!

laurenmarietta commented on jwql/website/apps/jwql/forms.py:
```diff
@@ -0,0 +1,158 @@
+"""Defines the forms for the ``jwql`` web app.
```
I was really excited by it, too! So clean and easy to use.

laurenmarietta commented on jwql/utils/utils.py:
```diff
@@ -117,3 +122,44 @@ def filename_parser(filename):
 
     return filename_dict
+
+def fileroot_parser(fileroot):
```
It is almost the same as `filename_parser` except it doesn't have the suffix and file extension included in the regex search - so that you can still get proposal/detector information/etc. from a string like `jw86600006001_02101_00008_guider1`rather than needing it to be `jw86600006001_02101_00008_guider1_cal.fits`. Do you think it's necessary?

If so, I'm thinking now it might make more sense to include it as another case within `filename_parser` rather than its own function...

bourque replied:
@laurenmarietta Yes, I think we should try making `filename_parser` more robust instead of having a second, mostly-similar function (which goes against the [DRY principle](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)).  I'm thinking it should be possible by adding some conditions on how one builds the string that is passed to `re.compile()` based on if the filename has `.fits` at the end, for example.

Perhaps we could also have a hard coded `FILETYPE_LIST` or something in `utils` that stores all of the valid filetypes (i.e. `uncal`, `cal`, `rate`, etc.).  It could also come in handy in this function.

laurenmarietta commented:
Okay, I think I have addressed all of your comments. I'm looking into how to write tests for `forms.py`, but it seems like there might be a steep learning curve, so I don't know if you would want me to try and figure that out before merging, or make that a separate PR.

bourque commented on jwql/utils/utils.py:
```diff
@@ -102,17 +103,23 @@ def filename_parser(filename):
     """
     filename = os.path.basename(filename)
 
+    file_root_name = (len(filename.split('.')) < 2)
```
@laurenmarietta Clever!

bourque commented:
@laurenmarietta Thanks for the changes, everything looks great and works well when I test it locally.  Unfortunately we won't yet be able to use this on dljwql.stsci.edu until I can fix the issues caused by merging #178.  

I agree that we can explore writing tests for `forms.py` in a separate issue/PR.
